modules/packages/OrderedSet/Treap.chpl

```diff
@@ -0,0 +1,862 @@
+/*
+ * Copyright 2004-2020 Hewlett Packard Enterprise Development LP
+ * Other additional copyright holders may be indicated within.
+ *
+ * The entirety of this work is licensed under the Apache License,
+ * Version 2.0 (the "License"); you may not use this file except
+ * in compliance with the License.
+ *
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+/*
+  This module contains a implementation of Treap,
+  which provides the functionality of OrderedSet.
+  Treap supporst insertion, deletion, query in O(lgN).
+
+  .. note::
+    Generally, users don't have to directly use this module. The methods of a treap 
+    are avaliable for an orderedSet. This page is for reference.
+*/
+module Treap {
+  import ChapelLocks;
+  private use HaltWrappers;
+  private use Sort;
+  private use Random;
+  private use IO;
+  private use Reflection;
+  private use OrderedSet only orderedSet;
+
+  pragma "no doc"
+  private param _sanityChecks = false;
+
+  // The locker is borrowed from List.chpl
+  // 
+  // We can change the lock type later. Use a spinlock for now.
+  //
+  pragma "no doc"
+  type _lockType = ChapelLocks.chpl_LocalSpinlock;
+
+  //
+  // Use a wrapper class to let heap methods have a const ref receiver even
+  // when `parSafe` is `true` and the orderedSet lock is used.
+  //
+  pragma "no doc"
+  class _LockWrapper {
+    var lock$ = new _lockType();
+
+    inline proc lock() {
+      lock$.lock();
+    }
+
+    inline proc unlock() {
+      lock$.unlock();
+    }
+  }
+
+  pragma "no doc"
+  var _treapRandomStream = new RandomStream(int);
+
+  /*
+    Helper procedure to get one random int
+  */
+  pragma "no doc"
+  proc _random(): int {
+    return _treapRandomStream.getNext();
+  }
+
+  pragma "no doc"
+  proc _checkType(type t) {
+    if isGenericType(t) {
+      compilerWarning('creating a orderedSet with element type ' + t:string, 2);
```

Comment:
Any specific reason for both warning and error?

Reply:
The compiler will only give out the first error message. So the first line needs to be a warning. Or they have to be combined into one error. Maybe not as readable as this way.

Reply:
If you mean code readability, single call to compilerError is much more readable. If you mean error readability, I understand the motivation to separate, but we have some pretty long error messages, and would like this to be a single compiler error message.

Reply:
I mean the error readability. Combined into a single error in https://github.com/chapel-lang/chapel/pull/16124/commits/45a6d99cf7ae1407cedf768bd3f6b2e051706c96